keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,153 @@
+3. Update quota code to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when evaluating usage for PVC.
+
+### Implementation
+
+We propose that by relaxing validation on PVC update to allow users to reduce `pvc.Spec.Resources`, it becomes possible to cancel previously issued expansion requests or retry expansion with a lower value if previous request has not been attemptd. In general - we know that volume plugins are designed to never peform actual shrinking of the volume, for both in-tree and CSI volumes. Moreover if a previously issued expansion has been successful and user
+reduces the PVC request size, for both CSI and in-tree plugins they are designed to return a successful response with NO-OP. So, reducing requested size will be a safe operation and will never result in data loss or actual shrinking of volume.
+
+We however do have a problem with quota calculation because if a previously issued expansion is successful but is not recorded(or partially recorded) in api-server and user reduces requested size of the PVC, then quota controller will assume it as actual shrinking of volume and reduce used storage size by the user(incorrectly). Since we know actual size of the volume only after performing expansion(either on node or controller), allowing quota to be reduced on PVC size reduction will allow an user to abuse the quota system.
+
+To solve aformentioned problem - we propose that, a new field will be added to PVC, called `pvc.Spec.AllocatedResources`. This field is only allowed to increase and will be set by the api-server to value in `pvc.Spec.Resources` as long as `pvc.Spec.Resources > pvc.Spec.AllocatedResources`.  Quota controller code will be updated to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when calculating usage of the PVC under questions. This does mean that - if user expanded PVC to a size that it failed to expand to and user retries expansion with lower value, the used quota will still use value reported by failed expansion request.
+
+In practice above limitation should be fine. We will add events to PV and PVC when we detect that user has reduced requested size. Admin can recover from this state either by manually increasing quota or by manually rebuilding the PVC object.
+
+#### User flow stories
+
+###### Case 1 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+
+
+###### Case 2 (controller+node expandable):
```

Comment:
Perhaps Case 3 would explain how `pvc.Spec.AllocatedResources` prevents user cheating:

1. User increases 10Gi PVC to 100Gi by changing `pvc.spec.resources.requests["storage"] = "100Gi"`
1. API server sets `pvc.spec.allocatedResources` to 100Gi.
1. Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota. (100Gi is the total space taken by the PVC)
1. Volume plugin starts slowly expanding to 100Gi. `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi until the resize is finished.
1. While the storage backend is resizing the volume, user requests size 20Gi by changing `pvc.spec.resources.requests["storage"] = "20Gi"`
1. Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
1. Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `100Gi`, as that's what the volume plugin did.

As result, `pvc.spec.resources.requests` is "20Gi", however, the user has access to "100Gi" volume. Since we used `pvc.Spec.AllocatedResources`, the total used space by the PVC for quota purposes is correct "100Gi".

\+ a note that everything would be super racy without the new field and would require controller-manager, admission-plugin and kubelet to use some sort of locking to prevent user from "shrinking" PVC while resize is in progress in the storage backend (i.e. in controller-manager or in kubelet).